osf/management/commands/set_institution_storage_regions.py

```diff
@@ -0,0 +1,114 @@
+
+    # Verify institution and region
+    try:
+        institution = Institution.objects.get(_id=institution_id)
+        region = Region.objects.get(_id=region_id)
+    except (Institution.DoesNotExist, Region.DoesNotExist) as e:
+        logger.error(f'Institution and/or Region not found: error={e}')
+        return
+    # Get or set region for institution
+    if region in institution.storage_regions.all():
+        logger.warning(f'Region [{region._id}] already set for Institution [{institution._id}]')
+        institution_storage_region = InstitutionStorageRegion.objects.get(
+            institution=institution,
+            storage_region=region
+        )
+        if institution_storage_region.is_preferred:
+            logger.warning(f'Region [{region._id}] already set as preferred for Institution [{institution._id}]')
+            return
+        if dry_run:
+            return
+    else:
+        if dry_run:
+            logger.warning(f'Dry-run: Region [{region._id}] has not been added to Institution [{institution._id}]')
+            return
+        else:
+            institution_storage_region = InstitutionStorageRegion.objects.create(
+                institution=institution,
+                storage_region=region
+            )
+            logger.info(f'Region [{region._id}] has been added to Institution [{institution._id}]')
+
+    # Make sure there is only one preferred region
+    preferred_regions = institution.storage_regions.filter(institutionstorageregion__is_preferred=True)
+    # Case 1: always set the region as preferred if there is no preferred region for the institution;
+    #         this executes even if the option `-p` / `--preferred` is not provided
+    if not preferred_regions:
+        institution_storage_region.is_preferred = True
+        institution_storage_region.save()
+        logger.info(f'Region [{region._id}] has been set as preferred choice for Institution [{institution._id}]')
+        return
+    # Case 2: do nothing and return if preferred region exists and if `is_preferred` is not set
+    if not is_preferred:
+        return
+    # Case 3: if `is_preferred` is set, clear existing preferred region(s) before set the new one;
+    #         use `filter()` instead of `get()` for two advantages: 1) `.update()` only works with
+    #         query set and 2) it clears extra preferred regions added unexpectedly without using
+    #         this command/script
+    for existing_preferred_region in preferred_regions:
+        InstitutionStorageRegion.objects.filter(
+            institution=institution,
+            storage_region=existing_preferred_region
+        ).update(is_preferred=False)
+    logger.info(f'Existing preferred regions have been cleared for Institution [{institution._id}]')
+    institution_storage_region.is_preferred = True
+    institution_storage_region.save()
+    logger.info(f'Region [{region._id}] has been set as preferred choice for Institution [{institution._id}]')
```

Comment:
For CR: Using this command to add allowed and preferred region(s) will eliminate the chance of creating multiple preferred regions. It also clears existing ones when setting a new preferred region.